pkg/apis/metrics/types.go

```diff
@@ -16,14 +16,205 @@ limitations under the License.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:"metadata"`
+	// Reference to the measured Node.
+	NodeName string `json:"nodeName"`
+	// Overall node metrics.
+	Total []AggregateSample `json:"total,omitempty" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+	// Metrics of system daemons tracked as raw containers, which may include:
+	//   "/kubelet", "/docker-daemon", "kube-proxy" - Tracks respective component metrics
+	//   "/system" - Tracks metrics of non-kubernetes and non-kernel processes (grouped together)
+	SystemContainers []RawContainerMetrics `json:"systemContainers,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// RawNodeMetricsList holds a list of RawNodeMetrics.
+// Endpoints which return this type respect unversioned.ListOptions
+type RawNodeMetricsList struct {
 	unversioned.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:"metadata,omitempty"`
+	// List of raw node metrics.
+	Items []RawNodeMetrics `json:"items"`
 }
 
-// Placeholder top-level pod resource metrics.
-type RawPod struct {
+// RawPod holds pod-level unprocessed sample metrics.
+type RawPodMetrics struct {
 	unversioned.TypeMeta `json:",inline"`
+	// Standard list metadata, since this is a synthetic resource.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:"metadata"`
+	// Reference to the measured Pod.
+	PodRef NonLocalObjectReference `json:"podRef"`
+	// Metrics of containers in the measured pod.
+	Containers []RawContainerMetrics `json:"containers" patchStrategy:"merge" patchMergeKey:"name"`
+	// Historical metric samples of pod-level resources.
+	Samples []PodSample `json:"samples" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+}
+
+// RawPodMetricsList holds a list of RawPodMetrics.
+// Endpoints which return this type respect unversioned.ListOptions
+type RawPodMetricsList struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:"metadata,omitempty"`
+	// List of raw pod metrics.
+	Items []RawPodMetrics `json:"items"`
+}
+
+// RawContainerMetrics holds container-level unprocessed sample metrics.
+type RawContainerMetrics struct {
+	// Reference to the measured container.
+	Name string `json:"name"`
+	// Metadata labels associated with this container (not Kubernetes labels).
+	// For example, docker labels.
+	Labels map[string]string `json:"labels,omitempty"`
+	// Historical metric samples gathered from the container.
+	Samples []ContainerSample `json:"samples" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+	// Application-specific metrics.
+	CustomMetrics []CustomMetric `json:"customMetrics,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
```

Comment:
No I'm not concerned with Heapster (or whatever is the source for HPA data) storing them.

My concern is that while running applications will be keeping metrics in memory in order to be scraped, by Prometheus or by cadvisor, cadvisor will also be keeping those metrics in memory to be scraped by Heapster (or other system). Obviously we need to ensure that memory footprint for anything running on nodes is kept to a minimum to get max resources available for applications. Anything we can do to achieve that we should do IMO.